promql/engine.go

```diff
@@ -576,45 +576,122 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *parser.Eval
 	return mat, warnings, nil
 }
 
-// subqueryOffsetRange returns the sum of offsets and ranges of all subqueries in the path.
-func (ng *Engine) subqueryOffsetRange(path []parser.Node) (time.Duration, time.Duration) {
+// subqueryTimes returns the sum of offsets and ranges of all subqueries in the path.
+// It also returns the min and max timestamp associated with the subquery.
+func (ng *Engine) subqueryTimes(path []parser.Node) (time.Duration, time.Duration, int64, int64) {
 	var (
-		subqOffset time.Duration
-		subqRange  time.Duration
+		subqOffset, subqRange      time.Duration
+		minTimestamp, maxTimestamp int64 = math.MaxInt64, math.MinInt64
 	)
 	for _, node := range path {
 		switch n := node.(type) {
 		case *parser.SubqueryExpr:
 			subqOffset += n.Offset
 			subqRange += n.Range
+			if n.Timestamp > 0 {
+				if n.Timestamp < minTimestamp {
+					minTimestamp = n.Timestamp
+				}
+				if n.Timestamp > maxTimestamp {
+					maxTimestamp = n.Timestamp
+				}
+			}
 		}
 	}
-	return subqOffset, subqRange
+	return subqOffset, subqRange, minTimestamp, maxTimestamp
 }
 
-func (ng *Engine) findMinTime(s *parser.EvalStmt) time.Time {
+func (ng *Engine) findMinMaxTime(s *parser.EvalStmt) (int64, int64) {
 	var maxOffset time.Duration
+	var minTimestamp, maxTimestamp int64 = math.MaxInt64, math.MinInt64
+
+	setMinMaxt := func(vsTs, subqMint, subqMaxt int64, vsOffset, vsRange, subqOffset, subqRange time.Duration) {
+		if vsTs == 0 {
+			// If we reached here, then subqMaxt is not math.MinInt64.
+			subqMaxt -= subqOffset.Milliseconds() + vsOffset.Milliseconds()
+			if subqMaxt > maxTimestamp {
+				maxTimestamp = subqMaxt
+			}
+			subqMint -= subqOffset.Milliseconds() + subqRange.Milliseconds()
+			subqMint -= vsOffset.Milliseconds() + vsRange.Milliseconds()
+			subqMint -= ng.lookbackDelta.Milliseconds()
+			if subqMint < minTimestamp {
+				minTimestamp = subqMint
+			}
+			return
+		}
+
+		// If the timestamp was set for vector/matrix selector, then
+		// the timestamp on the subquery does not matter.
+		vsTs -= vsOffset.Milliseconds()
+		if vsTs > maxTimestamp {
+			maxTimestamp = vsTs
+		}
+		if vsRange > 0 {
+			vsTs -= vsRange.Milliseconds()
+		} else {
+			vsTs -= ng.lookbackDelta.Milliseconds()
+		}
+		if vsTs < minTimestamp {
+			minTimestamp = vsTs
+		}
+	}
+
+	nodeWithNoTimestamp := false
+	wasAMatrixSelector := false
 	parser.Inspect(s.Expr, func(node parser.Node, path []parser.Node) error {
-		subqOffset, subqRange := ng.subqueryOffsetRange(path)
+		subqOffset, subqRange, subqMint, subqMaxt := ng.subqueryTimes(path)
 		switch n := node.(type) {
 		case *parser.VectorSelector:
+			if wasAMatrixSelector {
```

Comment:
This statefulness is a bit of a smell, why not run the code again? It's a noop, and keeps things simpler.

Reply:
We can actually ignore this but leads to incorrect min/max time and can end up selecting a bigger range (since with matrix selector we don't need a lookbackdelta and with vector selector we include it).

But since its a small time range, it should be fine I think and can adjust the test cases for that.

Reply:
The existence of wasAMatrixSelector is more where I see a potential issue here, from using elementTree back in the day this is a pattern you need to be quite careful with to get it right.

Reply:
I see that I have access to the `path`, I wont need `wasAMatrixSelector` now